lib/events/events.go

```diff
@@ -351,3 +353,47 @@ func Error(err error) *string {
 	str := err.Error()
 	return &str
 }
+
+// Wait for the given event to become available, or for a timeout, whichever is sooner.
+// Returns true if the event became available, or false if we timed out.
+func (s *bufferedSubscription) waitForNewEventWithTimeout(id int, timeout time.Duration) bool {
+	// Early exit, in the case where events are already available
+	s.mut.Lock()
+	cur := s.cur
+	s.mut.Unlock()
+	if id < cur {
+		return true
+	}
+
+	// From the point on, we know we're going to have to wait
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		s.mut.Lock()
+		if s.newEventNotification == nil {
+			s.newEventNotification = make(chan struct{})
+		}
+		newEventNotification := s.newEventNotification
+		s.mut.Unlock()
```

Comment:
I think there is a race here that will break it 

Cur=5
Id=5

T1: goes past the early exit
T2: emits new event, cur=6
T1: sets up notification channel at which point its too late
T1: times out without notifying about cur=6